icevision/tfms/albumentations/tfms.py

```diff
@@ -111,41 +113,170 @@ def apply(
+                get_transform(tfms_list, "OneOrOther") is None
+            ), " You must pass `crop_fn=None` to `aug_tfms` in case your dataset contains keypoints"
+
+            k = [xy for o in keypoints for xy in o.xy]
+            c = [label for o in keypoints for label in o.labels]
+            v = [visible for o in keypoints for visible in o.visible]
+            assert len(k) == len(c) == len(v)
+            params["keypoints"] = k
+            params["keypoints_labels"] = c
+
+            if get_transform(tfms_list, "Pad") is not None:
+                height_size, width_size = img.shape[:2]
+
+                t = get_transform(tfms_list, "SmallestMaxSize")
+                if t is not None:
+                    presize = t.max_size
+                    height_size, width_size = _func_max_size(
+                        height_size, width_size, presize, min
+                    )
+
+                t = get_transform(tfms_list, "LongestMaxSize")
+                if t is not None:
+                    size = t.max_size
+                    height_size, width_size = _func_max_size(
+                        height_size, width_size, size, max
+                    )
+
+        if masks is not None:
+            params["masks"] = list(masks.data)
 
         if bboxes is None:
             self.tfms.processors.pop("bboxes", None)
         if keypoints is None:
             self.tfms.processors.pop("keypoints", None)
 
-        if masks is not None:
-            params["masks"] = list(masks.data)
-
         d = self.tfms(**params)
 
         out = {"img": d["image"]}
         out["height"], out["width"], _ = out["img"].shape
 
         # We use the values in d['labels'] to get what was removed by the transform
+        if keypoints is not None:
+            tfms_kps = d["keypoints"]
+            # remove_invisible=False, therefore all points getting in are also getting out
+            assert len(tfms_kps) == len(k)
+            if get_transform(tfms_list, "Pad") is not None:
+                tfms_kps_n = filter_keypoints(tfms_kps, height_size, width_size, v)
+            else:
+                tfms_kps_n = filter_keypoints(tfms_kps, out["height"], out["width"], v)
+
+            l = list(chain.from_iterable(tfms_kps_n))
+            l = [
+                l[i : i + len(l) // len(keypoints)]
+                for i in range(0, len(l), len(l) // len(keypoints))
+            ]
+            assert len(l) == len(keypoints)
+            cl = keypoints[0].labels
+            # `if sum(k) > 0` prevents empty `KeyPoints` objects to be instantiated.
+            # E.g. `k = [0, 0, 0, 0, 0, 0]` is a flattened list of 2 points `(0, 0, 0)` and `(0, 0, 0)`. We don't want a `KeyPoints` object to be created on top of this list.
+            out["keypoints"] = [KeyPoints.from_xyv(k, cl) for k in l if sum(k) > 0]
         if labels is not None:
             out["labels"] = [labels[i] for i in d["labels"]]
+            if keypoints is not None:
```

Comment:
How can we refactor this so we don't need to do the check `if keypoints is not None:` in each object?

The logic that happens to filter the objects seems repetitive: `d["masks"][i] for i, k in zip(d["labels"], l) if sum(k) > 0`

Is it possible to somehow write a function that does this filtering and then call this function irrespective if `keypoints is None`?

Reply:
Yep makes sense. I will try to do that.

Reply:
I refactored the logic a bit. Is this sufficient?

Reply:
Resolving because it was addressed on discord =)